README.adoc
= Simple WebSocket Client
:revnumber: 0.2.2
:toc:
:sectnums:
:source-highlighter: prettify
ifndef::env-github[]
:icons: font
:icon-set: far
endif::[]
ifdef::env-github[]
:outfilesuffix: .adoc
:!toc-title:
:!sectnums:
:caution-caption: :fire:
:important-caption: :exclamation:
:note-caption: :paperclip:
:tip-caption: :bulb:
:warning-caption: :warning:
endif::[]

:proj-version: {revnumber}

image:https://img.shields.io/github/license/Axway-API-Management-Plus/ws-client-simple?style=plastic[]
image:https://img.shields.io/github/v/release/Axway-API-Management-Plus/ws-client-simple?style=plastic[]


== About the Tool

The tool is a simple client for WebSocket.
It simulates multiple client endpoints and sends/received messages to/from a WebSocket server.
It can be used to test WebSocket listeners of an API Gateway.
Check Axway API Gateway documentation to https://docs.axway.com/bundle/axway-open-docs/page/docs/apim_policydev/apigw_gw_instances/general_websocket/index.html[configure WebSocket connections].

== Usage

The default behavior is to send a single text message to the `ws://localhost:8025/echo` endpoint.
This is the default endpoint of the https://github.com/Axway-API-Management-Plus/ws-server-simple[simple WebSocket server].

[source,shell,subs="verbatim,attributes"]
----
$ java -jar ws-simple-client-{proj-version}-jar-with-dependencies.jar
INFO  10:29:06.452 - Connect to URI: ws://echo.websocket.org <1>
INFO  10:29:07.512 - [1]: connected <2>
INFO  10:29:07.513 - [1]: send - hello (#0) <3>
INFO  10:29:07.628 - [1]: received - hello (#0) <4>
INFO  10:29:09.522 - [1]: disconnected - 1 message(s) received <5>
INFO  10:29:09.528 - all endpoints closed
----
<1> Client connects to `ws://echo.websocket.org`
<2> The ID of the endpoint is shown in brackets.
Each simulated endpoint has its own identifier.
<3> Client endpoint sends first message to server.
<4> Client endpoint received first message from server
<5> WebSocket closed and client endpoint disconnected

To send two messages to the ws://192.168.56.201:8081/echo with three clients and delay of 500ms execute:

[source,shell,subs="verbatim,attributes"]
----
$ java -jar ws-simple-client-{proj-version}-jar-with-dependencies.jar --url ws://192.168.56.201:8081/echo --endpoints 3 --messages 2 --delay 500
INFO  10:33:07.377 - Connect to URI: ws://192.168.56.201:8081/echo
INFO  10:33:08.426 - [1]: connected <1>
INFO  10:33:08.672 - [2]: connected
INFO  10:33:08.890 - [3]: connected <2>
INFO  10:33:08.890 - [1]: send - hello (#0)
INFO  10:33:08.896 - [2]: send - hello (#0)
INFO  10:33:08.897 - [3]: send - hello (#0)
INFO  10:33:09.039 - [2]: received - hello (#0)
INFO  10:33:09.048 - [3]: received - hello (#0)
INFO  10:33:09.048 - [1]: received - hello (#0)
INFO  10:33:09.400 - [1]: send - hello (#1)
INFO  10:33:09.401 - [2]: send - hello (#1)
INFO  10:33:09.402 - [3]: send - hello (#1)
INFO  10:33:09.516 - [3]: received - hello (#1)
INFO  10:33:09.516 - [2]: received - hello (#1)
INFO  10:33:09.517 - [1]: received - hello (#1)
INFO  10:33:10.906 - [1]: disconnected - 2 message(s) received
INFO  10:33:10.907 - [2]: disconnected - 2 message(s) received
INFO  10:33:10.908 - [3]: disconnected - 2 message(s) received
INFO  10:33:10.912 - all endpoints closed
----
<1> First client endpoint connected to the server
<2> Up to three endpoints are simulate with this call.

== Installation

For installation just download the `ws-simple-client-{proj-version}-jar-with-dependencies.jar` file from the https://github.com/orgs/Axway-API-Management-Plus/packages?repo_name=ws-client-simple[GitHub repository] and execute it (see example below):

.Show help screen
[source,shell,subs="verbatim,attributes"]
----
$ java -jar ws-simple-client-{proj-version}-jar-with-dependencies.jar -h
Usage: ws-client-simple [options]
  Options:
    -h, --help
      Show help screen.
    --url
      Target WebSocket endpoint
      Default: ws://echo.websocket.org
    --endpoints
      Number of parallel client endpoints
      Default: 1
    --messages
      Number of messages to be send to the server
      Default: 1
    --delay
      Delay between sending messages (milliseconds)
      Default: 1000
----


== Contributing
Please read https://github.com/Axway-API-Management-Plus/Common/blob/master/Contributing.md[Contributing] for details on our code of conduct, and the process for submitting pull requests to us.

== Team
image:https://raw.githubusercontent.com/Axway-API-Management-Plus/Common/master/img/AxwayLogoSmall.png[Axway Logo] Axway Team

== License
link:./LICENSE[Apache License 2.0]


CHANGELOG.adoc
= Changelog

== Version 0.2.2
[cols="1,2,<10a", options="header"]
|===
|ID|Type|Description
|
|Enhancement
| * Change default URL to ws://localhost:8025/echo
* Upgrade to new version of log4j2
|===


== Version 0.1.1
[cols="1,2,<10a", options="header"]
|===
|ID|Type|Description
|https://github.com/Axway-API-Management-Plus/ws-client-simple/issues/1[#1]
|Enhancement
|In case of number of sent messages differ from to number of received message log an error instead of a warning.
|===

== Version 0.1.0
[cols="1,2,<10a", options="header"]
|===
|ID|Type|Description
|https://github.com/Axway-API-Management-Plus/ws-client-simple/issues/1[#1]
|Enhancement
|Add simple client for WebSockets.
|===


src/main/java/io/github/axway/apimplus/websocket/client/Endpoint.java
package io.github.axway.apimplus.websocket.client;

import java.io.IOException;
import java.net.URI;
import java.util.concurrent.atomic.AtomicInteger;

import jakarta.websocket.ClientEndpoint;
import jakarta.websocket.DeploymentException;
import jakarta.websocket.OnClose;
import jakarta.websocket.OnMessage;
import jakarta.websocket.OnOpen;
import jakarta.websocket.Session;
import jakarta.websocket.WebSocketContainer;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

@ClientEndpoint
public class Endpoint implements AutoCloseable {

	private static Logger log = LogManager.getLogger(Endpoint.class);

	private static AtomicInteger count = new AtomicInteger(0);

	private final int id;
	private final Session session;

	private volatile int msgCountReceived = 0;
	private volatile int msgCountSent = 0;

	public Endpoint(WebSocketContainer container, URI uri) throws DeploymentException, IOException {
		this.id = Endpoint.count.incrementAndGet();
		this.session = container.connectToServer(this, uri);
	}

	public int getId() {
		return this.id;
	}

	public int getReceivedMessageCount() {
		return this.msgCountReceived;
	}

	public int getSentMessageCount() {
		return this.msgCountSent;
	}

	public int getMissingMessageCount() {
		return this.msgCountSent - this.msgCountReceived;
	}

	public void sendText(String text) throws IOException {
		log("send - " + text);
		this.session.getBasicRemote().sendText(text);
		this.msgCountSent++;
	}

	public boolean isOpen() {
		return this.session != null && this.session.isOpen();
	}

	@Override
	public void close() throws IOException {
		if (this.session != null && this.session.isOpen()) {
			this.session.close();
		}
	}

	@OnMessage
	public void onMessage(String message, Session session) {
		this.msgCountReceived++;
		log("received - " + message);
	}

	@OnOpen
	public void onConnection(Session session) throws IOException {
		log("connected");
	}

	@OnClose
	public void onClose() {
		log("disconnected - " + this.msgCountReceived + " message(s) received");
	}

	private void log(String message) {
		log.info("[" + this.id + "]: " + message);
	}
}


src/main/java/io/github/axway/apimplus/websocket/client/WebSocketClient.java
package io.github.axway.apimplus.websocket.client;

import java.io.IOException;
import java.net.URI;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import jakarta.websocket.ContainerProvider;
import jakarta.websocket.DeploymentException;
import jakarta.websocket.WebSocketContainer;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import com.beust.jcommander.JCommander;
import com.beust.jcommander.ParameterException;

public class WebSocketClient implements AutoCloseable {

	private static final Logger log = LogManager.getLogger(WebSocketClient.class);

	private final Config config;
	private final WebSocketContainer container;
	private final List<Endpoint> endpoints = new ArrayList<Endpoint>();
	private final URI uri;

	public static void main(String[] args) throws Exception, IOException {
		int rc = 0;
		Config cfg = parseArgs(args);

		try (WebSocketClient client = new WebSocketClient(cfg)) {
			client.open();

			for (int m = 0; m < cfg.getMessageCount(); m++) {
				client.broadcast("hello (#" + m + ")");
				Thread.sleep(cfg.getDelayMillis());
			}

			client.waitForCompletion();

			if (!client.checkMessages()) {
				rc = 1;
			}
		} catch (Exception e) {
			log.error("error occurred", e);
			rc = 2;
		}

		System.exit(rc);
	}

	private static Config parseArgs(String[] args) {
		Config cfg = new Config();

		// Parse command line
		try {
			JCommander jc = JCommander.newBuilder().addObject(cfg).programName("ws-client-simple").build();
			jc.parse(args);
			if (cfg.isHelp()) {
				printHelpAndExit(jc);
			}
		} catch (ParameterException e) {
			log.error(e.getLocalizedMessage());
			System.exit(1);
		}

		return cfg;
	}

	private static void printHelpAndExit(JCommander jc) {
		jc.usage();
		System.exit(1);
	}

	public WebSocketClient(Config cfg) throws DeploymentException, IOException, InterruptedException {
		this.config = Objects.requireNonNull(cfg);
		this.container = ContainerProvider.getWebSocketContainer();
		this.uri = URI.create(this.config.getUrl());
	}

	public final void open() throws DeploymentException, IOException, InterruptedException {
		log.info("Connect to URI: " + this.uri);
		// create endpoints
		for (int i = 0; i < this.config.getEndpointCount(); i++) {
			this.endpoints.add(new Endpoint(this.container, this.uri));
			if (this.config.getDelayRampupMillis() > 0 && i < (this.config.getEndpointCount() - 1)) {
				Thread.sleep(this.config.getDelayRampupMillis());
			}
		}

		// check endpoints
		for (Endpoint ep : this.endpoints) {
			if (!ep.isOpen()) {
				close();
				throw new IOException("couldn't open all endpoints");
			}
		}
		log.debug("all endpoint available");
	}

	public void broadcast(String text) throws IOException {
		for (Endpoint ep : this.endpoints) {
			ep.sendText(text);
		}
	}

	public void waitForCompletion() throws IOException, InterruptedException {
		long untilTimeMillis = System.currentTimeMillis() + this.config.getTimeoutMillis();
		boolean completed = false;

		while (!completed && System.currentTimeMillis() < untilTimeMillis) {
			completed = true;
			Thread.sleep(500);

			for (Endpoint ep : this.endpoints) {
				int missingMessageCount = ep.getMissingMessageCount();
				if (missingMessageCount > 0) {
					completed = false;
					break;
				}
			}
		}

		if (!completed) {
			log.error("wait for completion canceled after timeout of " + this.config.getTimeoutMillis()
					+ " milliseconds");
		}
	}

	public boolean checkMessages() {
		boolean succeeded = true;
		for (Endpoint ep : this.endpoints) {
			int missingMessageCount = ep.getMissingMessageCount();
			if (missingMessageCount != 0) {
				succeeded = false;
				log.error(missingMessageCount + " messages are missing for endpoint [" + ep.getId() + "]");
			}
		}

		return succeeded;
	}

	@Override
	public void close() throws IOException {
		this.endpoints.forEach(ep -> {
			try {
				if (ep.isOpen())
					ep.close();
			} catch (Exception e) {
				log.error("failed to close endpoint", e);
			}
		});
		log.info("all endpoints closed");
	}
}


src/main/java/io/github/axway/apimplus/websocket/client/Config.java
package io.github.axway.apimplus.websocket.client;

import com.beust.jcommander.Parameter;

class Config {
	@Parameter(names = { "-h", "--help" }, description = "Show help screen.", help = true, order = 1)
	private boolean help = false;

	@Parameter(names = "--url", description = "Target WebSocket endpoint", order=2)
	private String url = "ws://localhost:8025/echo";

	@Parameter(names = "--endpoints", description = "Number of parallel client endpoints", order=3)
	private int endpointCount = 1;

	@Parameter(names = "--messages", description = "Number of messages to be send to the server", order=4)
	private int messageCount = 1;

	@Parameter(names = "--delay", description = "Delay between sending messages (milliseconds)", order=5)
	private long delayMillis = 1000;
	
	@Parameter(names = "--rampup", description = "Delay between ramp up a new client endpoint (milliseconds)", order=6) 
	private long delayRampupMillis = 0;
	
	@Parameter(names = "--timeout", description = "", order=7)
	private long timeoutMillis = 60000;


	public String getUrl() {
		return this.url;
	}

	public int getEndpointCount() {
		return this.endpointCount;
	}

	public int getMessageCount() {
		return this.messageCount;
	}

	public long getDelayMillis() {
		return this.delayMillis;
	}
	
	public long getDelayRampupMillis() {
		return this.delayRampupMillis;
	}
	
	public long getTimeoutMillis() {
		return this.timeoutMillis;
	}

	public boolean isHelp() {
		return this.help;
	}
}
